proposals/infra/INF-0004-validator-hashing.md

```diff
@@ -127,38 +127,107 @@ apply the hash. Instead it should apply the `PREVIEW` sentinel value from the
+    D3D12_BYTECODE_FLAG_VALIDATE_ALL_RELEASE_BYTECODE = 0x4,
+    D3D12_BYTECODE_FLAG_VALIDATE_ONLY_IF_DEBUG_LAYER_ENABLED_NO_FAILING = 0x8,
+    D3D12_BYTECODE_FLAG_SKIP_VALIDATION_IF_VALIDATOR_NOT_AVAILABLE = 0x10,
+    D3D12_BYTECODE_FLAGS_DEFAULT = 
+        D3D12_BYTECODE_FLAG_VALIDATE_BYTECODE_WITH_BYPASS_HASH | 
+        D3D12_BYTECODE_FLAG_VALIDATE_ONLY_IF_DEBUG_LAYER_ENABLED_NO_FAILING |
+        D3D12_BYTECODE_FLAG_SKIP_VALIDATION_IF_VALIDATOR_NOT_AVAILABLE
+} D3D12_BYTECODE_FLAGS;
+DEFINE_ENUM_FLAG_OPERATORS( D3D12_BYTECODE_VALIDATION_FLAGS )
+```
+
+Flag | Definition
+---|---
+`D3D12_BYTECODE_FLAG_VALIDATION_DISABLED` | Never invoke bytecode validation.
+`D3D12_BYTECODE_FLAG_VALIDATE_BYTECODE_WITH_BYPASS_HASH` | Only validate bytecode that has the `BYPASS` hash.
+`D3D12_BYTECODE_FLAG_VALIDATE_ALL_RELEASE_BYTECODE` | Validate all bytecode, hashed or `BYPASS`, except `PREVIEW`.  `PREVIEW` is excluded since that bytecode doesn't support validation at all.  Forcing validation for all release bytecode (hashed or `BYPASS`) could be useful to root out shaders that are hashed (implying validation ran on them), but there is an updated valdiator that catches issues that might have been missed.  Or the app might want validation on shaders that might have been hashed manually without the compiler's validator.
+`D3D12_BYTECODE_FLAG_VALIDATE_ONLY_IF_DEBUG_LAYER_ENABLED_NO_FAILING` | Only the debug layer performs bytecode validation.  If errors are found, it reports debug messages without failing the shader.  Accordingly if the debug layer isn't enabled no validation is done. Without this flag, the runtime validates bytecode and errors produce failure (e.g. shader creation failure), and if the debug layer is enabled it will also print a corresponding message.
+`D3D12_BYTECODE_FLAG_SKIP_VALIDATION_IF_VALIDATOR_NOT_AVAILABLE` | Regardless of how the other flags request validation, if the runtime or debug layer can't load `dxcompiler.dll` or `dxil.dll` and find a validation implementation they skip validation without failing.  If this flag is not set, the runtime will fail if it needs to validate bytecode and can't find a validator implementation.
+`D3D12_BYTECODE_FLAGS_DEFAULT` | By default validation is only done on bytecode with the `BYPASS` hash, only from the debug layer, and only if a validator implementation can be found.  Equivalent to the three flags: `D3D12_BYTECODE_FLAG_VALIDATE_BYTECODE_WITH_BYPASS_HASH`, `D3D12_BYTECODE_FLAG_VALIDATE_ONLY_IF_DEBUG_LAYER_ENABLED_NO_FAILING` and `D3D12_BYTECODE_FLAG_SKIP_VALIDATION_IF_VALIDATOR_NOT_AVAILABLE` |
+
+The first three flags above are mutually exclusive and one must be chosen.
+
+Flags are set via `ID3D12BytecodeOptions`:
+
+```C++
+[uuid(9291ad45-9a10-43a5-9c06-f5aa558e89a7), object, local, pointer_default(unique)]
+interface ID3D12BytecodeOptions
+    : IUnknown
+{
+    HRESULT SetBytecodeFlags(
+        [annotation("_In_")] D3D12_BYTECODE_FLAGS flags
+    );
```

Comment:
Should there be any other valdiation controls?  Like allowing behavior tweaks on the validator in some way? (e.g validator version setting?)

Reply:
Ok, so we need HLSL work item to add this no-hash option.

Reply:
I'm not sure I follow why we would add a no-hash option. Shouldn't we either run the validator and hash or not run the validator and thus not hash?

Reply:
This is for when the runtime calls the validator.  It's just interested in whether the bytecode is valid or not, no need to hash.

Reply:
I still question the need for this. Don't we have a flag to write in place? That should avoid increased memory impact, and if we're not writing the output we just wrote some memory that gets thrown away.

I'm unsure that the cost of calculating the hash for something that happens only in the debug layer is actually worth the extra support complexity of a new user exposed interface option.

Reply:
There is an  DxcValidatorFlags_InPlaceEdit flag.  The runtime could try to make sure it's only ever calling validate with it's own copy of bytecode, never passing in the APIs caller's memory.  But even if it did that, the runtime would lose track of what BYPASS the shader had originally since it would now be stomped by an unneeded hash.  Runtime could cache the bypass state before calling validator.  But it would seem cleanest overall if it could just ask for validation only without side effects.

Reply:
Caching the bypass state before calling the validator is less clean than calling the validator handling a potential error flag then needing to cache the bypass state anyways and call the validator without the flag?

It seems like whatever you need to do in the runtime to make this work with existing validators needs to be done regardless of whether or not we add a flag to the validator.

Reply:
For existing validators we'd just use the default behavior which is the validator gives us a copy of the shader that's hashed.

Much cleaner if we can first just try to validate with the new flag, which will avoid making a copy of the shader.  If the call fails E_INVALIDARG, since the new flag isn't recognized, we just call again without the flag and live with the fact that the validator made an unneeded copy of the shader.

Reply:
My earlier reply was exploring what we'd have to do if we used the in-place flag that you pointed out.  What we'd really do is above.  Easy way to avoid wasted shader copy if possible.  If the request isn't reasonable, then what I list as the fallback just has to be the only path.

Reply:
I think that if the fallback path needs to be implemented anyways, we should start there and only add the extra flag once we can measure the performance.

That allows us to understand the impact and make an informed decision about prioritization. Every hour we spend working on DXC is an hour we don't spend on Clang, and every new feature we add to DXC is a new feature we have to figure out how to support and/or deprecate as we move forward.

Reply:
> I'm unsure that the cost of calculating the hash for something that happens only in the debug layer is actually worth the extra support complexity of a new user exposed interface option.

Note: this will be called from normal runtime, not just from debug layer, when using `D3D12_BYTECODE_FLAG_VALIDATE_WHEN_HASH_BYPASSED` or `D3D12_BYTECODE_FLAG_VALIDATE_ALL_BYTECODE` flags.

Reply:
I can cut this flag from consideration - would you prefer that I just delete the section from the spec, or leave it in but labelled with "This section is cut until D3D team can prove that avoiding an unnecessary shader copy really matters."